### DISCUSSION TITLE
Make DataGrid editable/read-only on cell level

### DISCUSSION THREAD

**Discussion by ghost** [AvaloniaUI/Avalonia#10708] [Q&A] [answered]
I'm trying to figure out how to make only specific cells of a DataGrid editable/read-only depending on their value resp. some other value of the corresponding record.
Is that possible?

**Comment by timunie** [comment, marked as answer]
You can just cancel the attempt to edit a cell in `BeginningEdit-Event`:

```xml
<DataGrid Items="{Binding TestItems}" 
          AutoGenerateColumns="True" 
          IsReadOnly="False"
          BeginningEdit="DataGrid_OnBeginningEdit">
</DataGrid>
```

```cs
private void DataGrid_OnBeginningEdit(object? sender, DataGridBeginningEditEventArgs e)
{
    if (e.Row.DataContext is TestItem {Name: "Bla"})
    {
        e.Cancel = true;
    }
}
```

**Reply by ghost** [reply]
Thanks for the reply, I'll try that!
Btw how would I apply this to specific columns only?
I also have a requirement where I need to make only cells with the value "-" read-only and all the others editable.

**Reply by Mbucari** [reply]
> Btw how would I apply this to specific columns only?

`DataGridBeginningEditEventArgs` has a `Column` property that you can use to ignore columns you're not interested in. `DataGridColumn` has multiple properties you can use to identify the columns you're interested in, such as `Header`, `SortMemberPath`, and `DisplayIndex`.

**Reply by ghost** [reply]
Thanks, I got it working using DataGridBoundColumns Binding path.
One more question: How would you go about setting a different background on those specific cells?

**Reply by timunie** [reply]
That's a task for a Style. Look up original styles and use DevTools for checking what to style. A ocnverter could help.

**Reply by ghost** [reply]
I assume you mean using `<Style Selector="DataGridCell">`?
The problem is the DataContext I can bind to in the style is the row's DataContext, which means I can't bind to the cell's value. 
So I can change the background for all cells of a row, but how would I do it for specific cells only? (specifically only the ones that contain "-")

**Reply by timunie** [reply]
hm maybe a MultiConverter can help where you can pass the Column as parameter. Another idea is to use a `DataGridTemplateColumn` and apply it using the cell template. If you apply a CellEditingTemplate the cell can also be edited as any other cell.

**Reply by timunie** [reply]
Wait, there is also `CellStyleClasses`. So you can use classes for your selector.

**Reply by ghost** [reply]
I assume you mean passing `$parent[DataGridColumn]` to the MultiConverter? I guess that could work.
CellStyleClasses is not an AvaloniaProperty so I can't bind to it, otherwise I could use it to put a class on the cell depending on it's value.
DataGridTemplateColumn sounds like the best choice.

**Reply by timunie** [reply]
No, no. It's simple: 

```xml
<DataGrid.Columns>
 <DataGridTextColumn CellStyleClasses="myClass" />

...
<DataGrid.Styles>
  <Style Selector="DataGridCell.myClass>
    <Setter Property="Background" Value="{Binding ...}"
...
```

**Reply by ghost** [reply]
Ok so basically something like this (might also set IsEnabled as well instead of using the BeginningEdit event):
```xml
<DataGrid.Columns>
  <DataGridTextColumn CellStyleClasses="Column1" Binding="{Binding Column1}" />
  <DataGridTextColumn CellStyleClasses="Column2" Binding="{Binding Column2}" />

...
<DataGrid.Styles>
  <Style Selector="DataGridCell.Column1">
    <Setter Property="Background" Value="{Binding Column1, Converter=...}">
    <Setter Property="IsEnabled" Value="{Binding Column1, Converter=...}">
  </Style>
  <Style Selector="DataGridCell.Column2">
    <Setter Property="Background" Value="{Binding Column2, Converter=...}">
    <Setter Property="IsEnabled" Value="{Binding Column2, Converter=...}">
  </Style>
...
```

**Reply by timunie** [reply]
Yes, exactly. Does it work for you?

**Reply by ghost** [reply]
Yeah I think I can cover all my use cases with this, thanks!
